File: capgemini/batch-2/03-02-24/TwoUnique.py
arr = [2,3,23,5,6,654,32,1,2,3,23,1,5,6]


def getIthBit(n,i):
    n = n>>(i-1)
    return (n%2 == 1)

def getFirstSetBit(n):
    ans = 1
    while n%2 != 1:
        n = n>>1
        ans+=1
    return ans


def getUniqueNumbers(arr):
    allXor = 0
    for e in arr: allXor = allXor^e
    bitPosition = getFirstSetBit(allXor)
    seg1=[]
    seg2=[]
    for e in arr:
        if getIthBit(e,bitPosition) == True:
            seg1.append(e)
        else:
            seg2.append(e)
    ans1 = 0
    ans2 = 0
    for e in seg1: ans1=ans1^e
    for e in seg2: ans2=ans2^e

    return [ans1,ans2]


ans = getUniqueNumbers(arr)
print(ans)

File: JavaCrtRKGITM/22nd-august/TwoUnique.java

import java.util.*;

class TwoUnique{
    public static void main(String[] args){
        int[] arr = {3,2,6,7,24,95,6,95,7,3,2,9};
        int wholeXor = 0;
        for(int i = 0; i< arr.length; i++) wholeXor ^= arr[i];
        int fsbpos = getFirstSetBitPos(wholeXor);
        int ans1 = 0;
        int ans2 = 0;
        // List<Integer> seg1 = new ArrayList<>();
        // List<Integer> seg2 = new ArrayList<>();
        for(int i = 0; i< arr.length; i++){
            int ele = arr[i];
            if(getIthBit(ele,fsbpos) == 1){
                ans1^=ele;
            }else{
                ans2^=ele;
            }
        }
        
        // for(int i = 0 ; i < seg1.size(); i++) ans1 ^= seg1.get(i);
        // for(int i = 0 ; i < seg2.size(); i++) ans2 ^= seg2.get(i);
        System.out.print(ans1+" "+ans2);
    }
    public static int getFirstSetBitPos(int n){
        if(n == 0) return -1;
        int ans = 1;
        while(n%2 == 0){
            n = n>>1;
            ans+=1;
        }
        return ans;
    }
    public static int getIthBit(int n, int i){
        n = n>>(i-1);
        return n%2;
    }
}